[Constraint solver] Sink down initialization pattern handling

DougGregor opened the pull request:
Sink the constraint generation and solution application of initialization patterns, including all
of the logic for property wrappers, from the high-level entry point
`typeCheckBinding` down into the lower-level handling for
solution application targets.

DougGregor committed:
[Constraint system] Finish allowing one to record pattern types.

DougGregor committed:
[Constraint solver] Sink down initialization pattern constraint generation.

Sink the constraint generation for initialization patterns, including all
of the logic for property wrappers, from the high-level entry point
`typeCheckBinding` down into the lower-level constraint generation for
solution application targets.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -7267,6 +7267,10 @@ Optional<SolutionApplicationTarget> ConstraintSystem::applySolution(
       return None;
 
+    if (target.getExprContextualTypePurpose() == CTP_Initialization) {
+
```
This seems either incomplete or obsolete...

DougGregor replied:
Ah, it's incomplete. It's where the next bit of `typeCheckBinding` will be moving.

xedin reviewed (approved):
LGTM!

futurejones commented:
Can one of the admins verify this patch?

DougGregor committed:
Remove some unused code and fix a test

DougGregor committed:
[Constraint system] Remove typeCheckBinding’s ExprTypeCheckListener.

Sink the solution-application code from typeCheckBinding’s
ExprTypeCheckListener subclass into normal solution application, allowing
us to eliminate this subclass entirely. Down to one…

DougGregor committed:
[Constraint solver] Sink remaining logic from typeCheckBinding.

Move the remaining logic of typeCheckBinding, which consists of applying
the solution to the pattern, down into the normal solution application
path. typeCheckBinding() is now a thin shell over the normal
typeCheckExpression() convenience function.

DougGregor committed:
[Constraint solver] Make sure to reflect pattern updates after application.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test platform Linux

DougGregor commented:
@swift-ci please test platform Linux

DougGregor commented:
@swift-ci please test platform Linux

DougGregor commented:
@swift-ci please smoke test Linux
